Enhancing module

twerthi opened the pull request:
Added the ability to have all Access variables present, but allow for null or empty.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xsmbshare/20)
<!-- Reviewable:end -->

Commit:
Fixing parameter passing to New-SmbShare

Add checks for null Access parameters and removed from collection if they were null.

Commit:
Added removal of parameters that are null or emtpy

Added checks to see if Access parameters were present in the collection but were null or empty.  If they were null or empty, remove the variable from collection so when they are passed to New-SmbShare it wouldn't fail.

johlju commented:
Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---

*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 189 at r1](https://reviewable.io/reviews/powershell/xsmbshare/20#-LCdD9UG2HssN8mJ_OFK:-LCdD9UG2HssN8mJ_OFL:b-h15gq2) ([raw file](https://github.com/powershell/xsmbshare/blob/c5b4e2611385a3e9c772f2ea935f0742a662573e/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L189)):*
> ```PowerShell
>                 Write-Verbose "Parameter ChangeAccess is null or empty, removing from collection."
>                 # Remove the parameter
>                 $psboundparameters.Remove("ChangeAccess")
> ```

Curious why we need to do this. If the parameter is not passed to `Set-TargetResource` then is it not already not part of `$PSBoundParameters`? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)*
<!-- Sent from Reviewable.io -->

twerthi commented:
Holy cow, I did not see that anyone had looked at this!  My apologies.

If the parameters were empty or null, the setting was failing completely.  If they weren't part of the parameters being passed in, then it safely applied what was being requested.

johlju commented:
Maybe we should use [ValidateNotNullOrEmpty](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_advanced_parameters?view=powershell-6#validatenotnullorempty-validation-attribute) instead on the parameters? If we should not allow empty values, that should be changed on both the Set- and Test-TargetResource.

Would that work for the intended behavior?

johlju commented:
Also, is there a scenario where any of the parameters need to be assigned an empty value? So we don't do a breaking change.

twerthi commented:
The ValidateNotNullOrEmpty would prevent anyone passing in null or empty.  The scenario that I am using is

					xSmbShare "Share-$($Folder.Share.Name)"
					{
						Name = $Folder.Share.Name
						Ensure = "Present"
						FullAccess = $Folder.Share.FullAccess
						ReadAccess = $Folder.Share.ReadAccess
						ChangeAccess = $Folder.Share.ChangeAccess
						Path = $Folder.PhysicalPath
					}
In my data document, I may or may not have anything specified for ChangeAccess, for example, so that would be a null.  I wanted the data document to be flexible enough so that you don't have to specify all of the different levels, only the ones you wanted to change.  Does that make sense?

johlju commented:
Ah, I see what you are trying to do. I don't think this is the right approach. If a property should not be checked to be in a consistent state it should not be part of the configuration. This way if `ChangedAccess` on the target node would be anything other than an empty value the configuration would never become in a desired state. 
The configuration should be built so that only the properties that should be in desired state is present.

twerthi commented:
I think I understand, but I'm going to paraphrase to make sure we're on the same page.  Given the scenario where User A has ChangeAccess and we set ChangeAccess to null, the configuration won't do anything with ChangeAccess and User A will retain the permission.  Your point is that in this case, User A ChangeAccess should be removed, is that correct?

johlju commented:
I think User A's change access should be removed yes, because that is the desired state we want.

Let say we add a share 'TestShare', and we say the `ChangeAccess` should be `$null` (none). Then we want the desired state to be that no one should have change access.
If user A would be added manually to the share for change access. Then our configuration should remove user A, for the configuration to be in desired state.
If we want to keep user A that was added manually, we should not add the property `ChangeAccess` to the configuration. If the property is not present in the configuration the resource should not determine that properties state.

twerthi commented:
Gotcha.  You are correct, my solution does not currently handle that.  The removing of the parameters is still necessary because the SMS Share cmdlets will fail otherwise.  However, I need to change my solution so that users that have existing permissions that are not included in the specified desired state are added to the NoAccess collection, which will then remove their permissions.

Commit:
Update to MSFT_xSmbShare.psm1

Refactored permissions section, moved removal of permissions outside of collection if statements so that permissions are always removed and then re-added if the collections are not null.

twerthi commented:
That is incorrect, adding them to the NoAccess list only sets them to an explicit Deny versus removing them from the list.  I found that the original author checked to see if collection was null, if it wasn't, they removed all users with that access level and replaced it with the collection.  I have moved the removal out of the if statement and have it removing and then adding if the collection has anything in it.  This should ensure the correct desired state.

Commit:
Update MSFT_xSmbShare.psm1

Removed aliases ? and %, and replaced with spelled out operators.

johlju commented:
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @twerthi)

---

*[a discussion](https://reviewable.io/reviews/powershell/xsmbshare/20#-LFbb4uk8a5bp8WI8BWZ:-LFbb4uk8a5bp8WI8BW_:b-dao2qw) (no related file):*
There are currently no unit tests for this resource. Could you please add unit tests, for at least the functionality that this change affect?

The unit tests should be added to `\Tests\Unit`, and the unit test template is here https://github.com/PowerShell/DscResources/blob/master/Tests.Template/unit_template.ps1

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)*
<!-- Sent from Reviewable.io -->

twerthi commented:
I've not done these before, so it might take a little while as I work it into my schedule.  Thank you for providing the template that should help speed things up :)

Commit:
Created unit tests file

Not ready yet.

twerthi commented:
@johlju, I'm wondering if I can get a little guidance as I've not done this before.  The template calls a function Initialize-TestEnvironment, which isn't included in the script, may I assume it's a known method that creates a test VM of some sort?  If so, is there documentation that describes what this VM looks like in terms of drives so that I can perform the setup of a share to test with?

johlju commented:
@twerthi absolutely, happy too! 😃 Let me know if you have any further questions..

Tests are run using [AppVeyor](https://ci.appveyor.com/projects) - AppVeyor provides [build workers (VM's)](https://www.appveyor.com/docs/build-environment/#pre-installed-software) and if not specified, by using `image` keyword in the appveyor.yml file in the root of the repository, then the default build worker name is (for now) 'Visual Studio 2015'.
The build worker is a normal Windows Server 2012 R2, with a bunch of tools installed to help building and testing. But see it as a clean Windows Server 2012 R2 for this purpose.

Tests are started by appveyor.yml (AppVeyor control file) and tests is run under the [test framework DscResources.Tests](https://github.com/PowerShell/DscResource.Tests) which is cloned on run-time (even when you run tests locally with `Invoke-Pester`). `Initialize-TestEnvironment` is a helper function in the test environment. So everything in the header and footer of the test templates can safely be ignored, it just to get the prerequisite in place.
The test framework also has some common tests (for all resource modules) that needs to pass.

For example, you can find the test run for your PR if you click 'Details' on this (just below here).

![image](https://user-images.githubusercontent.com/7189721/42076020-b01d8380-7b73-11e8-9933-a6a01402b0be.png)

In  [AppVeyor](https://ci.appveyor.com/projects) you could create a free account (using your GitHub account) and can connect your fork to AppVeyor (add your fork as a project under Projects in AppVeyor). That way you can test your working branches, and send in PR's when you see they are passing tests. 

There are two types of tests - unit test and integration tests. **I will only expect you to write a unit test.** But if you want to give an integration tests a go it would be much appreciated. 
Unit test does not actually perform any actions, it just mocks the calls so it's possible to assert that the call mas wade. Integration tests actually performs the action, for example actually creates the share.  *Due to the limitations of AppVeyor some integration tests can not be done, for example it is not possible to restart the build worker during testing.*

**Example of unit tests for a resource**: https://github.com/PowerShell/SqlServerDsc/blob/dev/Tests/Unit/MSFT_SqlServerDatabaseMail.Tests.ps1

**Example of integration tests for a resource:** https://github.com/PowerShell/SqlServerDsc/blob/dev/Tests/Integration/MSFT_SqlServerDatabaseMail.Integration.Tests.ps1 **and the accompanying configuration** https://github.com/PowerShell/SqlServerDsc/blob/dev/Tests/Integration/MSFT_SqlServerDatabaseMail.config.ps1
 
I realized that we don't have that much information around this in DSC Resource Kit. But below is what I could find (created an issue to update the documentation).

More information around AppVeyor: https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#tests-in-appveyor

Test guidelines: https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md

Test framework (just for reference, normally only necessary read for new resource modules): https://github.com/PowerShell/DscResource.Tests

twerthi commented:
Ah!!  Thank you, this helps tremendously!

twerthi commented:
The pieces are finally fitting together in my head.  I've only been able to work on this sporadically so it wasn't quite clicking.  Spent some time on the 4th and I believe I understand now.  Sorry it's taking so long.

johlju commented:
@twerthi No worries, take your time. 🙂 If you get stuck, or need pointers, ping me and I try to help the best I can.

Commit:
Adding unit tests

Added unit tests for changed code.

twerthi commented:
@johlju would you mind looking at the AppVeyor results?  It's complaining that there are tab characters, but I cannot seem to find any.  It also failed because there wasn't a manifest file, what is that?

johlju commented:
Absolutely. I will do that as soon as I have time, tomorrow hopefully.

johlju commented:
Can you move the unit tests to the folder \Tests\Unit (create 'Unit')  and see if that solves the problem with the unit test? 
Regarding the tab-error, there must be some tab in the file MSFT_xSmbShare.psm1. Search for the tab-character and replace with four spaces instead.

If you still got error after that I will take a look again :)

Commit:
Moving unit test to another folder

Moved the unit tests to another folder

Commit:
Deleting unit tests file

Removing file, moved to another locaiton

Commit:
Replaced tabs with 4 spaces

Found tabs in file, replaced with 4 spaces

twerthi commented:
I appreciate the continued assistance.  I was able to get past those errors, but have encountered a new one.  From the error message, it looks like it may be absolutely expecting a { after some of the statements.  Typically I put these on the next line so that I can line them up for readability, do the curly braces have to be on the same line?  Or are there specific statements where this is a requirement?

johlju commented:
We should write the open brace on the next row, that’s also according to the style guideline. But, for Pester tests that is not possible due to how Pester is built. So for InModuleScope-, Describe-, Context- and It-blocks the open brace needs to be on the same line. :)

Commit:
Updated curly brace locations

Updated curly brace locations to conform to Pester rules.

twerthi commented:
Got past that, looks like I've got some more work to do.  It's past lunch so I have to get back to work, I'll see if I can get further tomorrow.

johlju reviewed (commented):
I think you are coming along great! Some suggestion while you continue resolving the unit test.

> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20#-:-LHhTG6j8If7mwAx6KvT:bdmaxdq)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @twerthi)

___
*[Tests/Unit/xSmbShare.tests.ps1, line 43 at r7](https://reviewable.io/reviews/powershell/xsmbshare/20#-LHhTVv582-ztHJTS9Dc:-LHhTVv582-ztHJTS9Dd:bg0zk64) ([raw file](https://github.com/powershell/xsmbshare/blob/b290ddc1a43600072385d70dfc735d789b8af46e/Tests/Unit/xSmbShare.tests.ps1#L43)):*
> ```PowerShell
> 
>     # Loop through collection and create user accounts
>     $userCollection | ForEach-Object {New-LocalUser -Name $_ -Description "Dummy account $_" -NoPassword}
> ```

Non-blocking: A unit tests should not actually change the test environment (that is an integration test). Is it possible to mock (fake) the users instead of actually creating them?
___
*[Tests/Unit/xSmbShare.tests.ps1, line 60 at r7](https://reviewable.io/reviews/powershell/xsmbshare/20#-LHhTzKY-rVskC9S8nGe:-LHhTzKY-rVskC9S8nGf:b-7jyy27) ([raw file](https://github.com/powershell/xsmbshare/blob/b290ddc1a43600072385d70dfc735d789b8af46e/Tests/Unit/xSmbShare.tests.ps1#L60)):*
> ```
> $mockUserAccounts = @()
> ```

Non-blocking: We could simplify this by

```
$mockUserAccounts = @(
    'User1'
    'User2'
    'User3'
    'User4'
    'User5'
)
```
___
*[Tests/Unit/xSmbShare.tests.ps1, line 155 at r7](https://reviewable.io/reviews/powershell/xsmbshare/20#-LHhV-Eu8Z8GZnG--kSu:-LHhV-Eu8Z8GZnG--kSv:b84nhef) ([raw file](https://github.com/powershell/xsmbshare/blob/b290ddc1a43600072385d70dfc735d789b8af46e/Tests/Unit/xSmbShare.tests.ps1#L155)):*
> ```
> -ModuleName $script:DSCResourceName
> ```

Non-blocking: I don't think you should need that parameter on the cmdlet, since you are in the InModuleScope-block? 
___
*[Tests/Unit/xSmbShare.tests.ps1, line 171 at r7](https://reviewable.io/reviews/powershell/xsmbshare/20#-LHhUT462l8bJCXMtwFB:-LHhUT462l8bJCXMtwFC:b-cyh63b) ([raw file](https://github.com/powershell/xsmbshare/blob/b290ddc1a43600072385d70dfc735d789b8af46e/Tests/Unit/xSmbShare.tests.ps1#L171)):*
> ```
> $testParameters = @{ChangeAccess = $mockDefaultChangeAccess}
> ```

Non-blocking: We could simplify this by

```
$testParameters = @{
    ChangeAccess = $mockDefaultChangeAccess
    ReadAccess = $mockDefaultReadAccess
    FullAccess = $mockDefaultFullAccess
    ...etc
}
```


<!-- Sent from Reviewable.io -->

twerthi committed:
Updating unit tests

twerthi committed:
Updated FolderEnumerationMode

FoderEnumerationMode for the actual object is an enum, but the Set and Test actions expect text, so I updated it to what it expected.

johlju commented:
Happy to help! 😃 Debugging your code showed that you tested apples and pears 🙂 

![image](https://user-images.githubusercontent.com/7189721/43306143-701b0376-917a-11e8-94e8-f14a7684ed7d.png)

First, the property returned from the `Get-TargetResource` is an array of strings, so we should check `$result.ChangeAccess[0]` (since the mock object we comparing to only have one value). Secondly, the row `$mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Change'}` returned an array of objects, because the `Where-Object` was never evaluated. If we use parenthesis around the block then `Where-Object` is evaluated, but then we return the mock object, which does not match the string returned by `$result.ChangeAccess[0]`. So if we instead change that part of the code to `($mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Change'}).AccountName` we will test apple against apple 🙂 

If you change the code to this, the `Get-TargetResource` test will pass.

```
$result.ChangeAccess[0] | Should Be ($mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Change'}).AccountName
$result.ReadAccess[0] | Should Be ($mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Read'}).AccountName
$result.FullAccess[0] | Should Be ($mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Full'}).AccountName
$result.NoAccess[0] | Should Be ($mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Deny'}).AccountName
```

johlju commented:
Also, I saw that there are no mock object for the  `$_.AccessRight -eq 'Deny'`, so this is essentially testing `result.NoAccess[0] | Should BeNullOrEmpty`

twerthi commented:
Ah!  Okay, I see what you're saying.

twerthi committed:
Fixing GetResource test

twerthi committed:
Removing NoAccess so it won't get bound

twerthi committed:
Removing module name reference

johlju commented:
@twerthi on the `Assert-MockCalled` you have `-ModuleName $script:DSCResourceName`, that is not neccessary as you already are in an `InModuleScope`-block.

johlju reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @twerthi)

___
*[Tests/Unit/xSmbShare.tests.ps1, line 115 at r9](https://reviewable.io/reviews/powershell/xsmbshare/20#-LIoNoLA4rux4i4WMaxG:-LIoNoLA4rux4i4WMaxH:b-ccmn4k) ([raw file](https://github.com/powershell/xsmbshare/blob/e5a247a9448dd7fea610ccdac7999e532e296ce0/Tests/Unit/xSmbShare.tests.ps1#L115)):*
> ```
> Mock Get-SmbShare
> ```

We should use named parameters, so this should be `Mock -CommandName Get-SmbShare`. Throughout all mocks.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 135 at r9](https://reviewable.io/reviews/powershell/xsmbshare/20#-LIoO3aM2ubTBlB4tJjl:-LIoO3aM2ubTBlB4tJjm:b1e2sp3) ([raw file](https://github.com/powershell/xsmbshare/blob/e5a247a9448dd7fea610ccdac7999e532e296ce0/Tests/Unit/xSmbShare.tests.ps1#L135)):*
> ```
> -Scope Context
> ```

We should change this to `-Scope It` and add/move the row `$result = Get-TargetResource @testParameters` into each `It`-block. Throughout the tests.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 139 at r9](https://reviewable.io/reviews/powershell/xsmbshare/20#-LIoOIuc3allYzs0nH69:-LIoOIuc3allYzs0nH6A:b-wcnb6c) ([raw file](https://github.com/powershell/xsmbshare/blob/e5a247a9448dd7fea610ccdac7999e532e296ce0/Tests/Unit/xSmbShare.tests.ps1#L139)):*
> ```
> Assert-MockCalled Get-SmbShareAccess -Exactly -Times 1 -Scope Context
> ```

Non-blocking: Optionally this `Assert-MockCalled` can be moved into the previous `It`-block, and change the description to `'Should call the correct mocks`. You can do similar throughout the tests.


<!-- Sent from Reviewable.io -->

johlju commented:
@twerthi You are doing an awesome job with these tests! Great work! 😃

twerthi commented:
Thanks!  Quick question regarding your feedback, you said to change the -Scope to It and move the $result = <operation> into each It block, would that mean the Mock statements move as well?

johlju commented:
The mock statements should normally not be in an It-block (there are exceptions). The mocks should be added in the `Context`-block, preferably in a `BeforeAll`-block or `BeforeEach`-block (which run before `It`-block).

*I saw that in the unit test template we don't have an example of where to normally place the mocks. I submit an issue for that so we can update the unit test template.*

twerthi commented:
Aha!  Yes, I see the problem now.  Thank you for the clarification.  Your help is most appreciated.  Almost there :)

twerthi committed:
Moved resource call, mock statements

Mock statements were not included in all context blocks, moved resource calls to each It block.

twerthi commented:
@johlju it is still saying that I did not declare a mock of Get-SmbShare for Set in the context of When the system is not in the desired state, but passing for context When the system is in the desired state.  In both instances the Mocks are set up exactly the same way within the context block, what am I missing?

johlju commented:
@twerthi If you referring to the `Assert-MockCalled` you should add `-Scope 'It'` to the calls so it only count's the mock calls inside the It-block.

twerthi committed:
Changed scoping for assert

Changed scoping for assert-mockcalled

twerthi commented:
<headsmack>Yeah, that makes sense ... feel kinda silly on that one :)

twerthi committed:
Fixing test

twerthi reviewed (commented):
I believe these conditions have been satisfied.

twerthi committed:
Fixing test target

twerthi committed:
Adding method to alter parameters

Adding method to alter the psboundparameters if anything is null or empty

twerthi committed:
Using method for Set resource

Moved code from inline of Set-Targetresource to using method Set-BoundParameters

twerthi commented:
@johlju made some progress today, down to 2 failing issues.  In the Test-TargetResource of the module, I see that it's calling Get-TargetResource for the current state.  I'm thinking I need to Mock this method call, but I'm not sure if the Get-TargetResource section of the unit test is in play or not.  Would you mind taking a look?

johlju commented:
Each function has a separate `Describe`-block so the describe block are not working together, but are separate instances with mocks and tests. Since you are testing the `Get-TargetResource` in one describe block, and you are testing that it returns the proper values, then you can just mock `Get-TargetResource` in the `Test-TargetResource` `Describe`-block to simplify the (number of) mocks.

twerthi committed:
Context were backwards

twerthi committed:
Fixed Test resource parameters

For the test portion, I had the wrong parameters assigned which returned true instead of false.

twerthi committed:
Fixed typo

twerthi committed:
Testing parameters

Purposefully changed parameter to make sure it will return false - I'm thinking I've done something wrong and I cannot find it.

twerthi commented:
I haven't forgotten about this, been bogged down by work.

twerthi committed:
Mocked Get-TargetResource

Test-TargetResource was calling the Get-TargetResource method.  I mocked the Get-TargetResource method so that it would return that the Test-TargetResource expected.

twerthi committed:
Mock of Get-TargetResource was incorrect

I incorrectly mocked the return of Get-TargetResource.  In addition, mocking the Get-TargetResource made it so it no longer called Get-SmbShare and Get-SmbShareAccess causing the test to fail

twerthi committed:
Because I'm dumb

Test-Target resource mock return values were set to the same values as testparameters, the not in desired state context would find no difference and return $true instead of $false.

twerthi committed:
Forced Ensure to present

Forced Ensure to present

twerthi commented:
@johlju The test for Test-TargetResource when not in desired state is failing and I cannot figure out why.  I mocked the Get-TargetResource method, setting the data shape to what is returned by the actual method.  I set the parameters to make sure that the list of users for the different access levels were different from the returned Get-TargetResource values.  It keeps returning true.  I've gone over the logic in the Test-TargetResource of the actual module and it should be returning false.  I'm stumped.

johlju reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 382 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW0v1X-a2GRwCG66rA:-LKW0v1X-a2GRwCG66rB:bbhw20z) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L382)):*
> ```PowerShell
>     
>     # Alter the bound parameters, removing anything that is null or emtpy
>     $PSBoundParameters = Set-BoundParameters -psboundparameters $PSBoundParameters
> ```

This function `Set-BoundParameters` are removing the properties you are trying to compare in your test.
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 395 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW15NK8-Y4-KrnXS-c:-LKW15NK8-Y4-KrnXS-d:b-zfxnqf) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L395)):*
> ```PowerShell
>         {
>             $Params = 'Name', 'Path', 'Description', 'ChangeAccess', 'ConcurrentUserLimit', 'EncryptData', 'FolderEnumerationMode', 'FullAccess', 'NoAccess', 'ReadAccess', 'Ensure'
>             if ($PSBoundParameters.Keys.Where({($_ -in $Params)}) | ForEach-Object {Compare-Object -ReferenceObject $PSBoundParameters.$_ -DifferenceObject $share.$_})
> ```

Can we write out a verbose message per property that is either in desired state or not in desired state?
___
*[Tests/Unit/xSmbShare.tests.ps1, line 49 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW0eJ96d7v7MESpm1M:-LKW0eJAAhdMHeIbDbqJ:b1l7f6z) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/Tests/Unit/xSmbShare.tests.ps1#L49)):*
<details><summary><i alt="summary">Quoted 18 lines of code…</i></summary>

> ```
>     # Define test user accounts
>     $mockUserAccounts = @(
>     "User1",
>     "User2",
>     "User3",
>     "User4",
>     "User5"
>     )
>  
>     # Declare mock objects
>     $mockChangeAccess = @("User1")
>     $mockReadAccess = @("User2")
>     $mockFullAccess = @("User3", "User5")
>     $mockNoAcess = @("User4")
>     $mockDefaultChangeAccess = @("User2")
>     $mockDefaultReadAccess = @("User3")
>     $mockDefaultFullAccess = @("User1")
>     $mockDefaultNoAccess = @()
> ```
</details>


THis must be inside the `InModuleScope`-block, otherwise they are not visible to the code in the `InModuleScope`-block


<!-- Sent from Reviewable.io -->

johlju commented:
@twerthi Let me know if my comments above helped you, or if can help you more 🙂

twerthi commented:
@johlju Will do.  I can only work on this as I have time, I'm bogged down at work at the moment, but will see if I can use my lunch hour soon.

johlju commented:
No worries. There is no hurry, just wanted to check if it helped you continue :) Take the time you need

twerthi committed:
Moved variables into InScope block

Variables were outside the InScope block and therefor not visible to the script block.

twerthi committed:
Attempting to provide more output

twerthi commented:
I'll need to spend some more time locally before pushing anything else.  Looks like something is definitely not working the way it should.

johlju commented:
@twerthi If you are not debugging with VSCode (and the PowerShell extension) then I highly suggest using VSCode (it can start a specific describe block) and stop at a break point, and then you can step through each line and see what the result is, and you will probably find the problem soon. 

![image](https://user-images.githubusercontent.com/7189721/44577123-8f883e00-a790-11e8-8786-5d206c86b28a.png)

twerthi committed:
Changed parameter name of method

I had used a reserved name for the input parameter of teh Set-BoundParameters function causing it to alter the $PSBoundParameters collection in an unexpected way.

twerthi committed:
Fixed tests

Finally got all tests running.

Commit:
Merge branch 'dev' into dev

twerthi commented:
Not sure why the build is failing, @johlju, can you see why?

twerthi commented:
@johlju Don't accept this yet, I think I may have missed something.

twerthi commented:
@johlju Never mind, it's there.

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @johlju and @twerthi)

*[a discussion](https://reviewable.io/reviews/powershell/xsmbshare/20#-LFbb4uk8a5bp8WI8BWZ:-LPlXm71EZrDLhabJ2a3:b-896fix) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

There are currently no unit tests for this resource. Could you please add unit tests, for at least the functionality that this change affect?

The unit tests should be added to `\Tests\Unit`, and the unit test template is here https://github.com/PowerShell/DscResources/blob/master/Tests.Template/unit_template.ps1
</blockquote></details>

Done.

---
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 382 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW0v1X-a2GRwCG66rA:-LPlXt3h1FXf3vl7Ue3_:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L382)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This function `Set-BoundParameters` are removing the properties you are trying to compare in your test.
</blockquote></details>

Done.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 135 at r9](https://reviewable.io/reviews/powershell/xsmbshare/20#-LIoO3aM2ubTBlB4tJjl:-LPlY5-45-Do2egcnIhR:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/e5a247a9448dd7fea610ccdac7999e532e296ce0/Tests/Unit/xSmbShare.tests.ps1#L135)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Scope Context
> ```

We should change this to `-Scope It` and add/move the row `$result = Get-TargetResource @testParameters` into each `It`-block. Throughout the tests.
</blockquote></details>

Done.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 49 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW0eJ96d7v7MESpm1M:-LPlY-Na19Klngdztt8c:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/Tests/Unit/xSmbShare.tests.ps1#L49)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>     # Define test user accounts
>     $mockUserAccounts = @(
>     "User1",
>     "User2",
>     "User3",
>     "User4",
>     "User5"
>     )
>  
>     # Declare mock objects
>     $mockChangeAccess = @("User1")
>     $mockReadAccess = @("User2")
>     $mockFullAccess = @("User3", "User5")
>     $mockNoAcess = @("User4")
>     $mockDefaultChangeAccess = @("User2")
>     $mockDefaultReadAccess = @("User3")
>     $mockDefaultFullAccess = @("User1")
>     $mockDefaultNoAccess = @()
> ```

THis must be inside the `InModuleScope`-block, otherwise they are not visible to the code in the `InModuleScope`-block
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi committed:
Added verbose flag

Added verbose flag to make sure the differences are seen.

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @johlju and @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 395 at r14](https://reviewable.io/reviews/powershell/xsmbshare/20#-LKW15NK8-Y4-KrnXS-c:-LPlaR317jQPmrgMuGGP:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/633dc5b089c0dfa0560efa93e74bfee3297b2009/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L395)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Can we write out a verbose message per property that is either in desired state or not in desired state?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi commented:
I thought I'd done everything necessary for this PR, did I miss something?

twerthi commented:
@johlju this has been marked as stale.  I know that you've only recently come back from vacation and this is probably one of many on your list.

johlju commented:
Yes there are a long backlog. :/ And have a lot of stuff to do at the day work currently, so can’t really catch up yet. 
But this and all others are not forgotten. :smile:

twerthi commented:
That make sense :)  Thanks for getting back to me so quickly.

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @johlju)

___
*[Tests/Unit/xSmbShare.tests.ps1, line 115 at r9](https://reviewable.io/reviews/powershell/xsmbshare/20#-LIoNoLA4rux4i4WMaxG:-LW-9otXBO2IvC-LfyOb:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/e5a247a9448dd7fea610ccdac7999e532e296ce0/Tests/Unit/xSmbShare.tests.ps1#L115)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> Mock Get-SmbShare
> ```

We should use named parameters, so this should be `Mock -CommandName Get-SmbShare`. Throughout all mocks.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi reviewed (commented):
> Reviewed 1 of 2 files at r16, 1 of 1 files at r17.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: all files reviewed, 6 unresolved discussions (waiting on @johlju)



<!-- Sent from Reviewable.io -->

twerthi commented:
How's this one coming?

johlju reviewed (commented):
Sorry for been away so long. It has been a lot of work the last quarter at the day job. 

Great job with the unit tests! I found a few review comments. This time I will review again faster. 

> Reviewed 1 of 2 files at r16, 1 of 1 files at r17.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20#-:-LXACRTA7Ff-exce4RVJ:b4saxg4)* status: all files reviewed, 5 unresolved discussions (waiting on @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 226 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXA9TGZ-O06_lnzs_5A:-LXA9TGZ-O06_lnzs_5B:b-kjdwhb) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L226)):*
> ```
> $psBoundParameters
> ```

Maybe we could use another variable name to assign the result here instead of using the built-in `$PSBoundParameters`variable? Then the built-in `$PSBoundParameters`variable is available for the other code if it is needed in the future.
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 382 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXA9zS-2dnKjfYbAHmB:-LXA9zS05ZtOJRBybP_g:b-a2zsfj) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L382)):*
> ```
> $PSBoundParameters
> ```

Maybe we could use another variable name to assign the result here instead of using the built-in `$PSBoundParameters`variable? Then the built-in `$PSBoundParameters`variable is available for the other code if it is needed in the future.
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 397 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAAIh_9-KhWkSCPSeR:-LXAAIh_9-KhWkSCPSeS:b-imw0fk) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L397)):*
> ```
> if ($PSBoundParameters.Keys.Where({($_ -in $Params)}) | ForEach-Object {$differences = Compare-Object -ReferenceObject $PSBoundParameters.$_ -DifferenceObject $share.$_; $differences})
> ```

Could we make the if-statement more readable by moving out some of the code before the if-block? 

https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#use-variables-rather-than-extensive-piping
___
*[Tests/Unit/xSmbShare.tests.ps1, line 40 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAAk1Z00sRkHuSuUZo:-LXAAk1Z00sRkHuSuUZp:b-khvvo7) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/Tests/Unit/xSmbShare.tests.ps1#L40)):*
> ```PowerShell
> 
>     InModuleScope $script:DSCResourceName {
>         # TODO: Optionally create any variables here for use by your tests
> ```

Please remove the commented rows that are from the template.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 174 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAB_Km687GGQhaRPqa:-LXAB_Kn9jhYfGt__75j:b-jll6aa) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/Tests/Unit/xSmbShare.tests.ps1#L174)):*
> ```PowerShell
> 
>                 # Set the script level parameters
>                 #$script:ChangeAccess = $mockSmbShareAccess | Where-Object {$_.AccessRight -eq 'Change'}
> ```

Please remove commented code if it is not needed.


<!-- Sent from Reviewable.io -->

twerthi committed:
Updated code with requested changes

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @johlju and @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 226 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXA9TGZ-O06_lnzs_5A:-LXLO5_q-xTzlIvHCEFv:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L226)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $psBoundParameters
> ```

Maybe we could use another variable name to assign the result here instead of using the built-in `$PSBoundParameters`variable? Then the built-in `$PSBoundParameters`variable is available for the other code if it is needed in the future.
</blockquote></details>

Done.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 40 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAAk1Z00sRkHuSuUZo:-LXLO7XiAE8ye8hHoBwB:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/Tests/Unit/xSmbShare.tests.ps1#L40)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please remove the commented rows that are from the template.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @johlju and @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 382 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXA9zS-2dnKjfYbAHmB:-LXLOI7o0WWtQwOIoeca:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L382)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $PSBoundParameters
> ```

Maybe we could use another variable name to assign the result here instead of using the built-in `$PSBoundParameters`variable? Then the built-in `$PSBoundParameters`variable is available for the other code if it is needed in the future.
</blockquote></details>

Done.
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 397 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAAIh_9-KhWkSCPSeR:-LXLOIjaA1sYLCPBZQok:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L397)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> if ($PSBoundParameters.Keys.Where({($_ -in $Params)}) | ForEach-Object {$differences = Compare-Object -ReferenceObject $PSBoundParameters.$_ -DifferenceObject $share.$_; $differences})
> ```

Could we make the if-statement more readable by moving out some of the code before the if-block? 

https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#use-variables-rather-than-extensive-piping
</blockquote></details>

Done.
___
*[Tests/Unit/xSmbShare.tests.ps1, line 174 at r17](https://reviewable.io/reviews/powershell/xsmbshare/20#-LXAB_Km687GGQhaRPqa:-LXLOCS47Nu1pjuE60w3:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/54793ecfbfbbd5f64a3bbff59f9b0a2c6f59f186/Tests/Unit/xSmbShare.tests.ps1#L174)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please remove commented code if it is not needed.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi commented:
I understand, been trying not to bug you too much since I can see you cover quite a large range of PRs.  Think I got all requested changes complete.

twerthi commented:
Is this one acceptable?

johlju reviewed (commented):
> Reviewed 2 of 2 files at r18.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: all files reviewed, 3 unresolved discussions (waiting on @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 402 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfpu28A5RpQEDbEtM:-LZKfpu28A5RpQEDbEtN:b-o8ptvy) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L402)):*
> ```
> Foreach
> ```

`foreach` (lower-case 'f')
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 408 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfxNyEWKZ-dHWfyvR:-LZKfxNyEWKZ-dHWfyvS:bdi8tem) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L408)):*
> ```
> if (($null -ne $differences) -and ($differences.Length -gt 0))
> ```

We should not need to enclose the evaluations in parenthesis? 

This should do? 

```
if ($null -ne $differences -and $differences.Length -gt 0)
```
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 410 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfTl66oV2pwUrp9mV:-LZKfTl66oV2pwUrp9mW:b-6q6ef1) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L410)):*
> ```
>  -Verbose
> ```

We should only output verbose messages when the user want it, when they do for example `Start-DscConfiguration -Verbose`. So it should not default to print the verbose messages. I think we can remove `-Verbose` here?


<!-- Sent from Reviewable.io -->

twerthi committed:
Updating upper case to lowercase on foreach

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @johlju)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 402 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfpu28A5RpQEDbEtM:-LZjsEIT2KHVi7XYYkBG:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L402)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> Foreach
> ```

`foreach` (lower-case 'f')
</blockquote></details>

Done.
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 408 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfxNyEWKZ-dHWfyvR:-LZjsORX4Ycpb-pwrdkz:b-gp1vlg) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L408)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> if (($null -ne $differences) -and ($differences.Length -gt 0))
> ```

We should not need to enclose the evaluations in parenthesis? 

This should do? 

```
if ($null -ne $differences -and $differences.Length -gt 0)
```
</blockquote></details>

I'm used to C# syntax, is this considered bad practice?
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 410 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfTl66oV2pwUrp9mV:-LZjszhnDnKP4jd3kwND:b-s6ztz9) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L410)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>  -Verbose
> ```

We should only output verbose messages when the user want it, when they do for example `Start-DscConfiguration -Verbose`. So it should not default to print the verbose messages. I think we can remove `-Verbose` here?
</blockquote></details>

In a previous comment, you'd asked that the differences be displayed.  I could not find another way to have the data be written to the DSC output from within a function call.  If there is another/better way to do this, please let me know and I'll update :)


<!-- Sent from Reviewable.io -->

johlju reviewed (commented):
> Reviewed 1 of 1 files at r19.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: all files reviewed, 1 unresolved discussion (waiting on @twerthi)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 408 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfxNyEWKZ-dHWfyvR:-L_6P7Oh9hZeOHx7J5sY:bw6o6o4) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L408)):*
<details><summary><i>Previously, twerthi (Shawn Sesna) wrote…</i></summary><blockquote>

I'm used to C# syntax, is this considered bad practice?
</blockquote></details>

It good as-is. 
___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 410 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfTl66oV2pwUrp9mV:-L_6PHC74yVm3-koF0ok:b-5syfvc) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L410)):*
<details><summary><i>Previously, twerthi (Shawn Sesna) wrote…</i></summary><blockquote>

In a previous comment, you'd asked that the differences be displayed.  I could not find another way to have the data be written to the DSC output from within a function call.  If there is another/better way to do this, please let me know and I'll update :)
</blockquote></details>

Sorry, let me explain better this time. It's correct to write out the differences with `Write-Verbose`, but the parameter `-Verbose` should not be used. The parameter `-Verbose` forces the verbose output regardless if the user wants verbose output or not.
If we remove the parameter `-Verbose` then the verbose output will be seen when the user asks for verbose output.
With my initial comment I meant that this should be written out, but only when the users asks for the output.

Also so now that it was missing the Message named parameter. SO if you change to this, then it's all good,

```
$differences | ForEach-Object { Write-Verbose -Message $_ }
```


<!-- Sent from Reviewable.io -->

twerthi committed:
Updating verbose message

Changed verbose to message.

twerthi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @johlju)

___
*[DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1, line 410 at r18](https://reviewable.io/reviews/powershell/xsmbshare/20#-LZKfTl66oV2pwUrp9mV:-L_8aphnCMqF0CdQdFjk:b-896fix) ([raw file](https://github.com/powershell/xsmbshare/blob/617d4e72b96bfdc6a72053ceb32fc3d3e0941eab/DscResources/MSFT_xSmbShare/MSFT_xSmbShare.psm1#L410)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Sorry, let me explain better this time. It's correct to write out the differences with `Write-Verbose`, but the parameter `-Verbose` should not be used. The parameter `-Verbose` forces the verbose output regardless if the user wants verbose output or not.
If we remove the parameter `-Verbose` then the verbose output will be seen when the user asks for verbose output.
With my initial comment I meant that this should be written out, but only when the users asks for the output.

Also so now that it was missing the Message named parameter. SO if you change to this, then it's all good,

```
$differences | ForEach-Object { Write-Verbose -Message $_ }
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

twerthi committed:
Syntax error in write-verbose call

I had the -Message backwards.

twerthi commented:
We all good now?

twerthi commented:
Checking in, anything that I'm missing?

gaelcolas commented:
Hi @twerthi,
Thanks for pinging us.
I'll see if @johlju as time for a final review or I'll ask him to fill me in and finish that off.

Do @ me tomorrow if it hasn't moved forward.

johlju commented:
On it.

johlju reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r20.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20#-:-Lau6tXt9wtP3DqBCZyV:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

johlju commented:
@gaelcolas can you merge this one as I have no permission to merge. :)

gaelcolas commented:
@twerthi can you add an entry in the changeling describing your change? I'll use it to merge the pr first thing tomorrow.

twerthi commented:
This project doesn't appear to have a CHANGELOG.md file in the root, do I need to add one or is it located somewhere else?

johlju commented:
@twerthi some repos have not split the change log into a seperate file. Then you find the unreleased section under the header changelog in the file README.md.

gaelcolas committed:
updated readme.md cosmetic

gaelcolas committed:
Added changelog entry

gaelcolas committed:
variable capitalization and cosmetic

gaelcolas reviewed (approved):
Hi @twerthi, I've made the update to the changelog and made some cosmetic changes (variable capitalization).
I'll merge this after the tests passes.
Thanks again for the PR!

gaelcolas reviewed (approved):
> Reviewed 1 of 2 files at r18, 2 of 2 files at r21.
> *[Reviewable](https://reviewable.io/reviews/powershell/xsmbshare/20)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

twerthi commented:
Oh!  I had fully planned on making the updates as requested, but got busy at work and am prepping for an event.  Thank you very much for taking care of that :)

gaelcolas commented:
No worries. Given we're all in different timezone I thought it'd be nice to close it before the monthly release tomorrow.
